### PR TITLE
common, osd: add fmtlib formatting of some OSD types

### PR DESCRIPTION
**Pull request by ronen-fr**
Adding fmtlib support for some "user-defined" OSD types.

A use-case added, as a proof of viability (per @tchaikov request).

Signed-off-by: Ronen Friedman <rfriedma@redhat.com>

### CODE REVIEW

File: src/common/hobject.h
Comment (tchaikov) [inline src/common/hobject.h:342]: this actually hurts the performance.

File: src/include/types.h
Comment (tchaikov) [inline src/include/types.h:240]: i'd suggest drop this change.

File: src/include/types.h
Comment (tchaikov) [inline src/include/types.h:258]: one empty line would suffice.

File: src/include/types.h
Comment (tchaikov) [inline src/include/types.h:267]: could you please put
```c++
const std::map<A, B, Comp, Alloc>& m
```
instead?

File: src/msg/msg_types.h
Comment (tchaikov) [inline src/msg/msg_types.h:22]: how about adding these specifier to a dedicated header file, like `msg_fmt.h`? so we don't have to include `fmt/format.h` header file everywhere?

File: src/osd/osd_types.h
Comment (tchaikov) [inline src/osd/osd_types.h:1013]: could you put `auto  format(eversion_t const& ev, FormatContext& ctx)` at a new line?

File: src/common/hobject.h
Comment (tchaikov) [inline src/common/hobject.h:338]: put `auto format(hobject_t const& ho, FormatContext& ctx)` at a new line.

File: src/CMakeLists.txt
Comment (tchaikov) [inline src/CMakeLists.txt:327]: why?

File: .gitmodules
Comment (tchaikov) [inline .gitmodules:52]: why?

File: src/common/hobject.h
Comment (ronen-fr) [reply src/common/hobject.h:342]: @tchaikov:
Agree on the performance, and will fix to include all the details
(these commits are just for you to OK the general idea and the change to the lib version).

Re using fmt/ostream directly: tried that, and immediately got into an endless loop during execution
(the ostream implementation tried to use fmt...). I couldn't find a clear way around (which may be the reason some
users advocate against using this option).

File: src/CMakeLists.txt
Comment (ronen-fr) [reply src/CMakeLists.txt:327]: 6.0 failed linking for some of the bins. The error seemed related to a bug that was worked around
in fmt 7.1.3. See the discussion in
https://github.com/fmtlib/fmt/issues/1753

File: .gitmodules
Comment (ronen-fr) [reply .gitmodules:52]: That's the part where I don't really understand what I am doing.
I need to bump up the version, but wasn't sure how to achieve that.

File: src/include/types.h
Comment (ronen-fr) [reply src/include/types.h:267]: a clear saying on east/west const debate :-) ?
OK

File: src/msg/msg_types.h
Comment (tchaikov) [reply src/msg/msg_types.h:22]: my concern is the build-time dependency. i am worried about the build time. as your change could have a profound impact of the amount of code to be build, as you are not only adding `fmt/format.h` to some very popular header files and also add this template specialization to it.

File: src/CMakeLists.txt
Comment (tchaikov) [reply src/CMakeLists.txt:327]: i'd suggest keep it as it is, unless we are using some new features or bug fixes which does not exist in 6.0. more importantly, quite a few developers are still using the GCC for compiling Ceph, AFAICT, and they don't have 7.1.3 prepackaged in the distro they are using: ubuntu focal ships 6.1.2, EPEL8 has 6.2.1.

if you really want to ensure that the clang users can be warned probably you could add something like
```cmake
if(CMAKE_CXX_COMPILER_ID STREQUAL "Clang" AND
    fmt_VERSION VERSION_LESS "7.1.3") 
  message(WARNING "fmt compiled with GCC does not link with Clang, see https://github.com/fmtlib/fmt/issues/1753")
endif()
```

please note, this issue does not impact the users who only use clang or GCC for compiling libfmt and the application using it. that's another reason why we should not enforce this version or bail out if the version is lower than 7.1.3 .

File: .gitmodules
Comment (tchaikov) [reply .gitmodules:52]: i created #41925 to partially address the linkage issue of clang + GCC.

Comment (ronen-fr) [reply .gitmodules:52]: Thanks

**Comment by ronen-fr** [comment]
All changes now in separate '.h' files. Thus -
a smaller, simpler, change, with much less effect on the rest of the system.

**Comment by ronen-fr** [comment]
jenkins test make check

**Review by tchaikov — APPROVED** [review]

**Comment by tchaikov** [comment]
@ronen-fr i don't feel comfortable when merging code which is not tested. would you please use at least some of the formatter in either crimson or classic osd ?

**Comment by ronen-fr** [comment]
> @ronen-fr i don't feel comfortable when merging code which is not tested. would you please use at least some of the formatter in either crimson or classic osd ?

@tchaikov Thanks. Will do. Was waiting for approval for this.

**Comment by ronen-fr** [comment]
Note: basically a compilation-tested change. Even more:
passed teu tests:

Runs:
original: yuriw-2021-09-09_21:20:54-rados-wip-yuri-testing-2021-09-09-1227-distro-basic-smithi (9F, 16D)
rerun: yuriw-2021-09-10_14:25:47-rados-wip-yuri-testing-2021-09-09-1227-distro-basic-smithi (4F, 1D)

Failures:

#6382926: “cephadm/test_dashboard_e2e.sh” - dashboard related failure. Seems to be https://tracker.ceph.com/issues/52417
#6382927: mgr/rook: might be https://tracker.ceph.com/issues/52321
#6382940: ceph_test_objectstore crash. Opened https://tracker.ceph.com/issues/52576
#6382942: seems the same as #6382926

The dead job: test issue (reimaging)

**Pull request merged**
